### PR TITLE
isync/mbsync: replace master/slave with far/near

### PR DESCRIPTION
**Pull request by KarlJoad**
### Description

Version 1.4.0 of isync/mbsync introduced a change where the words "master" and "slave" are no longer the main ways to specify the remote maildir and local maildir respectively.
Instead, the author changed to using "far" and "near" for the remote and local maildirs, respectively.

I changed the generated output configuration file to use `Far` and `Near` instead, and changed references to `master` or `slave` in supporting files to `far` and `near` instead.

**Please note that this PR will only be able to start working once [111907](https://github.com/NixOS/nixpkgs/pull/111907) is merged into nixpkgs' master branch.**

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by terlar** [comment]
Are these terms specifically for mbsync? Haven’t heard about this naming before, but mostly heard: primary/secondary

**Comment by KarlJoad** [comment]
Sort of, yes. For the configuration file, these are the terms specifically used by mbsync. Meaning the `.mbsyncrc` configuration file will use this naming scheme.

However, there is no reason the `.nix` files *have* to match that. In home-manager, we could use primary/secondary. Originally I used master/slave in the home-manager files because it translated better to the config file. That's the same reason I switched the names like I did here.

If you're looking for the specific commit where this change was introduced, it is: [c8f402e43f9c6a3c685fe0f716ffda741beeac13](https://sourceforge.net/p/isync/isync/ci/c8f402e43f9c6a3c685fe0f716ffda741beeac13/)

**Comment by KarlJoad** [comment]
I'm not sure the best way to approach backwards compatibility. mbsync will continue to support the old version for a while, while silently upgrading stateful files in a non-backwards compatible way. 
I'm not sure if I should use the `mkRenamedOptionModule` to ensure that people can still use the old naming (at least for a while).

**Comment by thiagokokada** [comment]
> I'm not sure if I should use the mkRenamedOptionModule to ensure that people can still use the old naming (at least for a while).

I think this is the preferred approach, specially if they're equivalent.

It will show a warning in red for anyone that doesn't change, so I think it would make people migrate fast.

**Comment by rycee** [comment]
Yeah, definitely use `mkRenamedOptionModule`.

**Comment by KarlJoad** [comment]
The only reason I can think of to *not* use `mkRenamedOptionModule` is that when switching to `Far` and `Near` in the config, the state files for the mailbox are silently upgraded in-place. So, even when someone uses a `master`/`slave` variable in their home-manager config, if we switch out `Master` for `Far` in the generated configuration file, the mailbox is not backwards-compatible. Meaning if someone reverts to an earlier generated config file that uses `Master` (after using `Far`), mbsync will not work.

I found this out by running a quick test on my own machine, and I ended up having to delete all the downloaded mail and redownload it.

**Comment by thiagokokada** [comment]
> The only reason I can think of to _not_ use `mkRenamedOptionModule` is that when switching to `Far` and `Near` in the config, the state files for the mailbox are silently upgraded in-place. So, even when someone uses a `master`/`slave` variable in their home-manager config, if we switch out `Master` for `Far` in the generated configuration file, the mailbox is not backwards-compatible. Meaning if someone reverts to an earlier generated config file that uses `Master` (after using `Far`), mbsync will not work.
> 
> I found this out by running a quick test on my own machine, and I ended up having to delete all the downloaded mail and redownload it.

I think adding a note about this in Release Notes should be sufficient, but maybe @rycee has a better idea.

**Comment by berbiche** [comment]
> the mailbox is not backwards-compatible

@KarlJoad this is indeed a big deal-breaker to using `mkRenamedOptionModule`.

I'm in favor of not using it and adding a warning if the options `master` and `slave` are used with an explanation and link to the changes.

**Comment by rycee** [comment]
Ah, right. So at the moment you start using the near/far in the mbsync config it will convert the database format, otherwise it will keep the database format unchanged?

If so, then perhaps

1. changing the master/slave option names to correspond with the new near/far ones with `mkRenamedOptionModule` and
2. write the near/far naming in the mbsync configuration for stateVersion ≥ 21.05, otherwise write use the master/slave naming in the configuration file.
3. Add a note in the state version change section of the release notes.

**Comment by sivizius** [comment]
Current state? At least the generation of the mbsyncrc configuration-file should use Far/Near instead of Master/Slave, this would be a non-breaking change.

**Comment by KarlJoad** [comment]
Sorry about the long delay. School picked up in a major way.

That should be a non-breaking change on `home-manager`'s end, and I believe I have already done that. However, this is a change that will affect people who might downgrade.

Right now,  the breakage is due to the checks for the old Master/Slave always returning `true`. I need to dig into that some more. I just haven't had the time recently.

**Comment by KarlJoad** [comment]
Figured out the problem. Accidentally left a "farPattern" testing string in the function that checks all accounts. All fixed now, and everything is passing.

**Review by berbiche — DISMISSED** [review]
LGTM!

@rycee I'll let you merge :smile:

**Comment by KarlJoad** [comment]
It may look good to merge, but it doesn't quite do what I want yet. It does not support backwards-compatibility yet. Right now, it will simply fail to evaluate if you use the older `masterPattern` and `slavePattern`. I am still trying to figure out how to support them while throwing an error that they will be deprecated.

**Comment by berbiche** [comment]
> It may look good to merge, but it doesn't quite do what I want yet. It does not support backwards-compatibility yet. Right now, it will simply fail to evaluate if you use the older `masterPattern` and `slavePattern`. I am still trying to figure out how to support them while throwing an error that they will be deprecated.

You could  use `warnings` under `config` and check whether any of these patterns are used

**Comment by KarlJoad** [comment]
> You could use `warnings` under `config` and check whether any of these patterns are used

That is what I am currently doing now. However, @rycee suggested:
>     2. write the near/far naming in the mbsync configuration for stateVersion ≥ 21.05, otherwise write use the master/slave naming in the configuration file, and
> 
>     3. add a note in the state version change section of the release notes.

I have not gotten around to getting the right version of the config file written out depending on the `stateVersion`. I forget right now, but there was something else I wanted to fix before a merge.

Not to mention I need to resolve a merge conflict in the `news.nix` file.

**Comment by KarlJoad** [comment]
@rycee I believe this should finish up the work required for mbsync. Please take a look over it. I believe I have finished everything you requested, including sending out `warning` messages when someone uses the old `masterPattern`/`slavePattern` setup, a set of tests for this new behavior, and made note of this in `news.nix`.

**Review by sumnerevans — APPROVED** [review]
a couple of nits, but otherwise looks good!

**Comment by teto** [comment]
Sorry for the conflict with news.nix do you mind rebasing while adressing the nitpicks, I would like to merge.

File: modules/programs/mbsync.nix
Comment (KarlJoad) [reply modules/programs/mbsync.nix]: Just switched to `(c: c)`. If that fails to test properly, then I will change it back. I feel the `anyTrue` function might be more effective as a library function, because I think it would be useful in more places than this.

**Comment by KarlJoad** [comment]
@teto `news.nix` resolved and most nitpicks handled.

**Review by sumnerevans — APPROVED** [review]

File: modules/misc/news.nix
Comment (rycee) [inline modules/misc/news.nix:2035]: Shouldn't this be conditioned to only show for users of the mbsync module? I guess people who don't wouldn't care about this.

**Pull request merged**
